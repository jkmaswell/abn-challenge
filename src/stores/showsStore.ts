import { ShowModel } from '@/models/ShowModel'
import type { Show } from '@/types/show'
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'

export const useTvShowsStore = defineStore('tvShows', () => {
  // State
  const shows = ref<Show[]>([])
  const searchQuery = ref('')

  // Actions
  const loadShows = async () => {
    const allShows = await ShowModel.fetchShows()

    shows.value = sortShowsByRating(allShows)
  }

  // Getters
  const genres = computed(() => {
    const allGenres = shows.value.flatMap(show => show.genres)
    
    return [...new Set(allGenres)]
  })

  const showsByGenre = computed(() => (genre: string) => sortShowsByRating(shows.value.filter(show => show.genres.includes(genre))))

  const filteredShows = computed(() => sortShowsByRating(
    shows.value.filter(show =>
      show.name.toLowerCase().includes(searchQuery.value.toLowerCase()),
    ),
  ))

  const sortShowsByRating = (shows: Show[]): Show[] => {
    return [...shows].sort((a, b) => (b.rating.average || 0) - (a.rating.average || 0))
  }

  return {
    shows,
    searchQuery,
    loadShows,
    genres,
    showsByGenre,
    filteredShows,
  }
})

export type TvShowsStore = ReturnType<typeof useTvShowsStore>;