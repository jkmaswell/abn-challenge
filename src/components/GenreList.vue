<script setup lang="ts">
import { useTvShowsStore } from '@/stores/showsStore'
import { storeToRefs } from 'pinia'
import { onMounted } from 'vue'
import ShowCard from './ShowCard.vue'

// Stores
const store = useTvShowsStore()
const { genres } = storeToRefs(store)

// Methods
const showsByGenre = (genre: string) => store.showsByGenre(genre)

// Hooks
onMounted(async () => {
  await store.loadShows()
})
</script>

<template>
  <div class="genre-list">
    <div
      v-for="genre in genres"
      :key="genre"
      class="genre-list__genre"
    >
      <h2 class="genre-list__title">
        {{ genre }}
      </h2>
      <div class="genre-list__shows">
        <ShowCard
          v-for="show in showsByGenre(genre)"
          :key="show.id"
          :show="show"
        />
      </div>
    </div>
  </div>
</template>

<style scoped lang="scss">
.genre-list {
  &__genre {
    margin-block-end: 2rem;
  }

  &__title {
    font-size: 2rem;
    margin-block-end: 1.5rem;
  }

  &__shows {
    display: flex;
    flex-wrap: nowrap;
    gap: 1.5rem;
    overflow-x: auto;
    padding-block: 0.5rem;
    padding-inline: 0;
  }
}
</style>