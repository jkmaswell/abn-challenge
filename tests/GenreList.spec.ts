import GenreList from '@/components/GenreList.vue'
import { TvShowsStore, useTvShowsStore } from '@/stores/showsStore'
import { flushPromises, mount } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { RouterLinkMock } from './mocks/RouterLinkMock'
import { ShowCardMock } from './mocks/ShowCardMock'

describe('GenreList.vue', () => {
  let wrapper: ReturnType<typeof mount>
  let store: TvShowsStore

  beforeEach(async () => {
    setActivePinia(createPinia())

    store = useTvShowsStore() as TvShowsStore

    store.shows = [
      { id: 1, name: 'Show 1', genres: ['Drama'], rating: { average: 7.5 }, image: { original: '', medium: '' }, summary: '' },
    ]

    vi.spyOn(store, 'loadShows')

    wrapper = mount(GenreList, {
      global: {
        components: {
          RouterLink: RouterLinkMock,
        },
      },
    })
    await flushPromises()
  })

  it('renders genres correctly', () => {
    const genreTitles = wrapper.findAll('.genre-list__title').map(el => el.text())

    expect(genreTitles).toEqual(['Drama'])
  })

  it('renders ShowCard components for each show in a genre', () => {
    const showCards = wrapper.findAllComponents(ShowCardMock)

    expect(showCards.length).toBe(1)  // Only one show in the 'Drama' genre
    expect(showCards[0].props('show').name).toBe('Show 1')
  })

  it('calls loadShows on mount', () => {
    expect(store.loadShows).toHaveBeenCalled()
  })

  it('does not render ShowCard components for genres with no shows', () => {
    const comedySection = wrapper.findAllComponents(ShowCardMock).filter(card => card.props('show').genres.includes('Comedy'))

    expect(comedySection.length).toBe(0)  // No shows for 'Comedy' genre
  })
})