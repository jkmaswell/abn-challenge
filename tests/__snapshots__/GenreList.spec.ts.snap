// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GenreList.vue > matches snapshot 1`] = `"<div data-v-47bca01b="" class="genre-list"></div>"`;
